File: Tracker/src/main/java/com/cyrillrx/tracker/consumer/ScheduledConsumer.java
package com.cyrillrx.tracker.consumer;

import com.cyrillrx.tracker.TrackerChild;
import com.cyrillrx.tracker.event.TrackEvent;
import com.cyrillrx.tracker.utils.Utils;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.TimeUnit;

/**
 * @author Cyril Leroux
 *         Created on 21/04/16.
 */
public class ScheduledConsumer extends EventConsumer<Queue<TrackEvent>> {

    private TimeUnit timeUnit;
    private long timeDuration;

    public ScheduledConsumer(TrackerChild tracker, Queue<TrackEvent> queue, TimeUnit unit, long duration) {
        super(tracker, queue);
        timeUnit = unit;
        timeDuration = duration;
    }

    @Override
    public void run() {

        running = true;
        while (running) {
            consume();
            Utils.wait(timeDuration, timeUnit);
        }
    }

    protected void consume() {

        // Events to be sent
        final List<TrackEvent> eventBucket = new ArrayList<>();

        TrackEvent event;
        while (!events.isEmpty()) {
            event = events.poll();

            // If stop received, stop the batch and send pending events
            if (STOP_EVENT.equals(event)) {
                running = false;
                break;
            }
            eventBucket.add(event);
        }

        try {
            doConsume(eventBucket);

        } catch (Exception e) {
            // TODO implement retry policy
            // TODO log warning
            events.addAll(eventBucket);
        }
    }

    private synchronized void doConsume(List<TrackEvent> events) { tracker.track(events); }
}

File: tracker/src/main/java/com/cyrillrx/tracker/consumer/ScheduledConsumer.kt
package com.cyrillrx.tracker.consumer

import com.cyrillrx.logger.Logger
import com.cyrillrx.tracker.event.TrackEvent
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import java.util.ArrayList

/**
 * @author Cyril Leroux
 *         Created on 21/04/2016.
 */
abstract class ScheduledConsumer(
    pendingEvents: Channel<TrackEvent>,
    private val intervalMillis: Long
) :
    EventConsumer(pendingEvents) {

    override suspend fun consume() {

        val eventBatch = events.getAsBatch()

        try {
            doConsume(eventBatch)

        } catch (cause: Exception) {
            Logger.info(TAG, "Error while consuming. Re-injecting events the event queue.", cause)
            eventBatch.forEach { events.send(it) }
        }

        delay(intervalMillis)
    }

    protected abstract fun doConsume(events: List<TrackEvent>)

    private fun Channel<TrackEvent>.getAsBatch(): ArrayList<TrackEvent> {

        val eventBatch = ArrayList<TrackEvent>()

        var event: TrackEvent

        while (!isEmpty) {

            event = poll() ?: break

            // If stop received, stop the batch and send pending events
            if (event == STOP_EVENT) {
                running = false
                break
            }
            eventBatch.add(event)
        }

        return eventBatch
    }

    companion object {
        private val TAG = ScheduledConsumer::class.java.simpleName
    }
}